server/api/cvrs.py

```diff
@@ -782,20 +793,98 @@ def join_ballots_to_interpretations(
+    # 2. Expect either [ another ZIP file ] or [ another ZIP file and a CSV ].
+    # 3. If a CSV is found, parse it as a scanned ballot information CSV.
+    # 4. Unzip the CVRs ZIP file.
+    # 5. Parse out the contest and choice names. We have to do this in a separate pass since our
+    #    storage scheme for interpretations requires knowing all of the contest and choice names up
+    #    front.
+    # 3. Then, parse out the interpretations.
+
+    wrapper_zip_file = retrieve_file(jurisdiction.cvr_file.storage_path)
+    files = unzip_files(wrapper_zip_file)
+    wrapper_zip_file.close()
+
+    cvrs_zip_file: Union[BinaryIO, None] = None
+    scanned_ballot_information_file: Union[BinaryIO, None] = None
+    for name, file in files.items():
+        if name.lower().endswith(".zip"):
+            cvrs_zip_file = file
+        if name.lower().endswith(".csv"):
+            scanned_ballot_information_file = file
+
+    if cvrs_zip_file is None:
+        raise UserError("Expected a file with a .zip extension.")
+
+    def construct_cvr_guid_to_unique_identifier_mapping(
+        scanned_ballot_information_file: BinaryIO,
+    ) -> Dict[str, str]:
+        validate_not_empty(scanned_ballot_information_file)
+        text_file = decode_csv(scanned_ballot_information_file)
+
+        # Skip #FormatVersion row
+        first_line = text_file.readline()
+        if "#FormatVersion" not in first_line:
+            raise UserError(
+                "Expected first line of scanned ballot information CSV to contain '#FormatVersion'."
+            )
+        validate_comma_delimited(text_file)
+        # validate_comma_delimited resets the cursor to the start of the file so skip the
+        # #FormatVersion row again
+        text_file.readline()
+        scanned_ballot_information_csv: CSVIterator = csv.reader(
+            text_file, delimiter=","
+        )
+        scanned_ballot_information_csv = reject_no_rows(scanned_ballot_information_csv)
+
+        headers_row = next(scanned_ballot_information_csv)
+        if len(headers_row) > 0:
+            headers_row[0] = headers_row[0].lstrip("#")
+        header_indices = get_header_indices(headers_row)
+
+        cvr_guid_to_unique_identifier_mapping: Dict[str, str] = {}
+        for i, row in enumerate(scanned_ballot_information_csv):
+            row_number = (
+                i + 3
+            )  # Account for zero indexing, #FormatVersion row, and header row
+            cvr_guid = column_value(
+                row,
+                "CvrId",
+                row_number,
+                header_indices,
+                file_name="scanned ballot information CSV",
+            )
+            unique_identifier = column_value(
+                row,
+                "UniqueIdentifier",
+                row_number,
+                header_indices,
+                file_name="scanned ballot information CSV",
+            )
+            if cvr_guid != "" and unique_identifier != "":
```

Comment:
empty strings are handled as errors by `column_value` unless you set `required = False`